### PR TITLE
Create wallpaper portal

### PR DESCRIPTION
**Pull request by danirabbit**
This replaces [the contract](https://github.com/elementary/switchboard-plug-pantheon-shell/tree/main/set-wallpaper-contract) we currently ship

### CODE REVIEW

File: src/Wallpaper/Portal.vala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        ObjectPath handle,
+        string app_id,
+        string parent_window,
+        string uri,
+        HashTable<string, Variant> options
+    ) throws DBusError, IOError {
+        var set_on = "both"; // Possible values are background, lockscreen or both.
+        var show_preview = false;
+
+        if ("set-on" in options && options["set-on"].get_type_string () == "s") {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:25]: This style allows 0 memory allocation and a single lookup on the GLib.HashTable :wink: 
```suggestion
        unowned var set_on = options.get ("set-on");
        if (set_on != null && set_on.is_of_type (GLib.VariantType.STRING)) {
```

File: src/Wallpaper/Portal.vala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                folder.make_directory_with_parents ();
+            } catch (Error e) {
+                warning (e.message);
+            }
+        }
+
+        return folder;
+    }
+
+    private string get_local_bg_directory () {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:108]:
```suggestion
    private static string get_local_bg_directory () {
```

File: src/Wallpaper/Portal.vala [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        return Path.build_filename (Environment.get_user_data_dir (), "backgrounds") + "/";
+    }
+
+    private string[] get_system_bg_directories () {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:112]:
```suggestion
    private static string[] get_system_bg_directories () {
```

File: src/Wallpaper/Portal.vala [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        if (background_directories.length == 0) {
+            warning ("No background directories found");
+        }
+
+        return background_directories;
+    }
+
+    private File ensure_local_bg_exists () {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:95]:
```suggestion
    private static File ensure_local_bg_exists () {
```

File: src/Wallpaper/Portal.vala [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            dest = ensure_local_bg_exists ().get_child (filename);
+            source.copy (dest, FileCopyFlags.OVERWRITE | FileCopyFlags.ALL_METADATA);
+        } catch (Error e) {
+            warning (e.message);
+        }
+
+        return dest;
+    }
+
+    private string[] get_bg_directories () {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:78]:
```suggestion
    private static string[] get_bg_directories () {
```

File: src/Wallpaper/Portal.vala [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        foreach (unowned string directory in get_bg_directories ()) {
+            if (path.has_prefix (directory)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private File? copy_for_library (File source) {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:63]:
```suggestion
    private static File? copy_for_library (File source) {
```

File: src/Wallpaper/Portal.vala [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+        if (file != null) {
+            var settings = new Settings ("org.gnome.desktop.background");
+            settings.set_string ("picture-uri", file.get_uri ());
+        }
+
+        return 0;
+    }
+
+    private bool get_is_file_in_bg_dir (File file) {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:51]:
```suggestion
    private static bool get_is_file_in_bg_dir (File file) {
```

File: src/Wallpaper/Portal.vala [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    [DBus (name = "SetWallpaperURI")]
+    public async uint set_wallpaper_uri (
+        ObjectPath handle,
+        string app_id,
+        string parent_window,
+        string uri,
+        HashTable<string, Variant> options
+    ) throws DBusError, IOError {
+        var set_on = "both"; // Possible values are background, lockscreen or both.
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:22]:
```suggestion
        unowned string set_on = "both"; // Possible values are background, lockscreen or both.
```

File: src/Wallpaper/Portal.vala [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        string app_id,
+        string parent_window,
+        string uri,
+        HashTable<string, Variant> options
+    ) throws DBusError, IOError {
+        unowned string set_on = "both"; // Possible values are background, lockscreen or both.
+        unowned bool show_preview = false; // FIXME: implement preview dialog
+
+        unowned var _set_on = options.get ("set-on");
+        if (_set_on.get_type_string () == "s") {
```
Comment (tintou) [inline src/Wallpaper/Portal.vala:26]: I insist, but this means no string allocation ;)
```suggestion
        if (_set_on.is_of_type (GLib.VariantType.STRING)) {
```

Comment (danirabbit) [reply src/Wallpaper/Portal.vala]: Oh that's way cleaner, thanks!

**Review by tintou — APPROVED** [review]
Let's do it!

**Pull request merged**